src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -1859,23 +1870,10 @@ public override void List(CriteriaImpl criteria, IList results)
 
 				string[] implementors = Factory.GetImplementors(criteria.EntityOrClassName);
 				int size = implementors.Length;
-
```

Comment:
I reviewed the code once more. It will require some public API changes on ISessionFactoryImplementor level.
Shall I continue to refactor rest of the code?